## Dataset was all clearly labelled with no null values. Random emails to training set but not to the test_data

In [ ]:
import json
import pandas as pd
import torch

In [4]:
def load_json_to_dataframe(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        data = [json.loads(line) for line in f]
    
    return pd.DataFrame(data)

def count_entity_tags(df):
    tag_counts = Counter()
    for tags in df["ner_tags"]:
        tag_counts.update(tags)
    
    return {"B-EMAIL": tag_counts.get("B-EMAIL", 0), "B-PER": tag_counts.get("B-PER", 0), "I-PER": tag_counts.get("I-PER", 0)}


In [5]:
file_path = "email_added_data.json"  
df = load_json_to_dataframe(file_path)
print("Loaded DataFrame:")
df.head()

Loaded DataFrame:


sequence  \
0  Since joe.montana@company.com then , only Terr...   
1  He was portrayed alfredhitchcock@company.com b...   
2  The sharptooth_random65@yahoo.com egg eventual...   
3  In the video Kelis is walking down a street in...   
4  According to food writer Sharon Tyler Herbst ,...   

                                              tokens  \
0  [Since, joe.montana@company.com, then, ,, only...   
1  [He, was, portrayed, alfredhitchcock@company.c...   
2  [The, sharptooth_random65@yahoo.com, egg, even...   
3  [In, the, video, Kelis, is, walking, down, a, ...   
4  [According, to, food, writer, Sharon, Tyler, H...   

                                            ner_tags  
0  [O, B-EMAIL, O, O, O, B-PER, I-PER, O, O, O, O...  
1  [O, O, O, B-EMAIL, O, B-PER, I-PER, O, O, O, O...  
2        [O, B-EMAIL, O, O, O, O, O, O, O, B-PER, O]  
3  [O, O, O, B-PER, O, O, O, O, O, O, O, O, O, O,...  
4  [O, O, O, O, B-PER, I-PER, I-PER, O, O, O, O, ...

In [4]:
# person token count same, email count added
entity_counts = count_entity_tags(df)
print("Entity Counts:")
print(entity_counts) 

Entity Counts:
{'B-EMAIL': 20076, 'B-PER': 40264, 'I-PER': 29466}


In [5]:
def count_rows_missing_tags(df):
    missing_b_email = df[~df["ner_tags"].apply(lambda tags: "B-EMAIL" in tags)].shape[0]
    missing_b_per = df[~df["ner_tags"].apply(lambda tags: "B-PER" in tags)].shape[0]
    missing_i_per = df[~df["ner_tags"].apply(lambda tags: "I-PER" in tags)].shape[0]
    
    return {"Rows missing B-EMAIL": missing_b_email, "Rows missing B-PER": missing_b_per, "Rows missing I-PER": missing_i_per}

missing_tag_counts = count_rows_missing_tags(df)
print("Rows Missing Entity Tags:")
print(missing_tag_counts)


Rows Missing Entity Tags:
{'Rows missing B-EMAIL': 14334, 'Rows missing B-PER': 3, 'Rows missing I-PER': 10363}


## this shows out of 28,516 rows, 28,516 - 14,334 = 14,182 (roughly 50%) have email added to them in random manner.
